materialize-kafka/src/binding_info.rs

```diff
@@ -0,0 +1,331 @@
+
+    let res = http
+        .post(&url)
+        .basic_auth(username, Some(&password))
+        .json(&serde_json::json!({
+            "schema": schema_str,
+            "schemaType": "AVRO"
+        }))
+        .send()
+        .await?;
+
+    if !res.status().is_success() {
+        let status = res.status();
+        let body = res.text().await?;
+        anyhow::bail!(
+            "request POST {} failed with status {}: {}",
+            url,
+            status,
+            body
+        );
+    }
+
+    let res: SchemaId = res.json().await?;
+    Ok(res.id)
+}
+
+struct ComputedBinding {
+    topic: String,
+    key_schema: avro::Schema,
+    schema: avro::Schema,
+    key_ptr: Vec<Pointer>,
+    field_names: Vec<String>,
+}
+
+fn binding_info(binding: &Binding) -> Result<ComputedBinding> {
+    let field_selection = binding.field_selection.to_owned().unwrap();
+    let mut fields: Vec<String> = field_selection
+        .keys
+        .iter()
+        .chain(field_selection.values.clone().iter())
+        .map(|f| f.to_owned())
+        .collect();
+    if !field_selection.document.is_empty() {
+        fields.push(field_selection.document);
+    }
+
+    let collection = binding.collection.to_owned().unwrap();
+
+    let collection_schema = if collection.read_schema_json.is_empty() {
+        collection.write_schema_json
+    } else {
+        collection.read_schema_json
+    };
+
+    let json_schema = doc::validation::build_bundle(&collection_schema)?;
+    let validator = doc::Validator::new(json_schema)?;
+    let collection_shape = Shape::infer(&validator.schemas()[0], validator.schema_index());
+    let collection_locations = collection_shape.locations();
+    let collection_projections = collection.projections;
+
+    // The fields in the schema must be sorted to match the ordering of
+    // constructed values for Avro serialization to work.
+    let mut fields_sorted = fields.clone();
+    fields_sorted.sort();
+
+    // Create a Shape based on the selected fields and their projections which
+    // will be used to generate an appropriate schema.
+    let mut shape = Shape::nothing();
+    shape.type_ = schema::types::OBJECT;
+    shape.object.properties = fields_sorted
+        .iter()
+        .map(|field| {
+            let (shape, is_required) =
+                field_to_shape(field, &collection_projections, &collection_locations);
+
+            ObjProperty {
+                name: field.clone().into(),
+                is_required,
+                shape,
+            }
+        })
+        .collect();
+
+    let key_ptr = field_selection
+        .keys
+        .iter()
+        .map(|k| {
+            // The selected key fields are treated as unescaped strings. For
+            // example, a key field like `some/nested~1key` is taken directly as
+            // a property named "some/nested~1key" with that pointer rather than
+            // unescaping the ~1 into a nested pointer of ("some", "nested/key").
```

Comment:
If you have a document with a structure like this:

```json
{
  "some": {
    "nested/key": "value"
  } 
}
```

You can select the "nested/key" field, which is nested in the containing "some" object, and Flow provides a flattened alias for that field that is the string "some/nested~1key" - the "nesting" is represented by slashes, and actual slashes in the field names are escaped as `~1`.

SQL materializations etc. take that alias directly and create a column for it. We want to do the same thing with Kafka, creating a top-level property if a user explicitly selects a nested field (nested fields are not recommended by default, since the full structure of all top-level fields is materialized by default, so there's not really any reason to select nested fields, but you could if you want to).

The use of `Token::from_str` is in contrast to using `Pointer::from_str` more directly since `Pointer::from_str` would interpret such a string as representing a nested structure rather than just treating it as the desired named of the field.

It's all pretty obscure and not likely to come up very often, but it was an edge case I uncovered in my testing.